src/System.Diagnostics.EventLog/src/System/Diagnostics/SharedUtils.cs

```diff
@@ -0,0 +1,335 @@
+        }
+
+        internal static void EnterMutex(string name, ref Mutex mutex)
+        {
+            string mutexName = null;
+            if (CurrentEnvironment == W2kEnvironment)
+                mutexName = "Global\\" + name;
+            else
+                mutexName = name;
+
+            EnterMutexWithoutGlobal(mutexName, ref mutex);
+        }
+
+        internal static void EnterMutexWithoutGlobal(string mutexName, ref Mutex mutex)
+        {
+            bool createdNew;
+            MutexSecurity sec = new MutexSecurity();
+            SecurityIdentifier everyoneSid = new SecurityIdentifier(WellKnownSidType.AuthenticatedUserSid, null);
+            //sec.AddAccessRule(new MutexAccessRule(everyoneSid, MutexRights.Synchronize | MutexRights.Modify, AccessControlType.Allow));
+            //Mutex tmpMutex = new Mutex(false, mutexName, out createdNew, sec);
+            Mutex tmpMutex = new Mutex(false, mutexName, out createdNew);
+            SafeWaitForMutex(tmpMutex, ref mutex);
+        }
+
+        private static bool SafeWaitForMutex(Mutex mutexIn, ref Mutex mutexOut)
+        {
+            Debug.Assert(mutexOut == null, "You must pass in a null ref Mutex");
+            while (true)
+            {
+                if (!SafeWaitForMutexOnce(mutexIn, ref mutexOut))
+                    return false;
+                if (mutexOut != null)
+                    return true;
+                Thread.Sleep(0);
+            }
+        }
+
+        [MethodImplAttribute(MethodImplOptions.NoInlining)]
+        private static bool SafeWaitForMutexOnce(Mutex mutexIn, ref Mutex mutexOut)
+        {
+            bool ret;
+            RuntimeHelpers.PrepareConstrainedRegions();
+            try
+            { }
+            finally
+            {
+                Thread.BeginCriticalRegion();
+                Thread.BeginThreadAffinity();
+                int result = WaitForSingleObjectDontCallThis(mutexIn.SafeWaitHandle, 500);
+                switch (result)
+                {
+                    case NativeMethods.WAIT_OBJECT_0:
+                    case NativeMethods.WAIT_ABANDONED:
+                        mutexOut = mutexIn;
+                        ret = true;
+                        break;
+
+                    case NativeMethods.WAIT_TIMEOUT:
+                        ret = true;
+                        break;
+
+                    default:
+                        ret = false;
+                        break;
+                }
+                if (mutexOut == null)
+                {
+                    Thread.EndThreadAffinity();
+                    Thread.EndCriticalRegion();
+                }
+            }
+
+            return ret;
+        }
+
+        [DllImport(Interop.Libraries.Kernel32, ExactSpelling = true, SetLastError = true, EntryPoint = "WaitForSingleObject")]
+        private static extern int WaitForSingleObjectDontCallThis(SafeWaitHandle handle, int timeout);
```

Comment:
@danmosemsft again a "DontCallThis" function ;)

Reply:
You might follow the recommendation provided here: https://github.com/dotnet/corefx/pull/24335#discussion_r142190411. You both want to use the same method and therefore should put it into common. Whoever will merge last into master will need to rebase.

Reply:
same one?

Reply:
Yes it's the same API